plutus-ir/src/Language/PlutusIR/Analysis/Dependencies.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE LambdaCase       #-}
+-- | Functions for computing the dependency graph of variables within a term or type.
+module Language.PlutusIR.Analysis.Dependencies (Node (..), DepGraph, runTermDeps, runTypeDeps) where
+
+import qualified Language.PlutusCore               as PLC
+import qualified Language.PlutusCore.Name          as PLC
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Analysis.Usages as Usages
+
+import           Control.Monad.Reader
+
+import qualified Algebra.Graph.Class               as G
+import           Data.Coerce
+import qualified Data.Set                          as Set
+import           Lens.Micro
+
+-- | A node in a dependency graph. Either a specific 'PLC.Unique', or a specific
+-- node indicating the root of the graph. We need the root node because when computing the
+-- dependency graph of, say, a term, there will not be a binding for the term itself which
+-- we can use to represent it in the graph.
+data Node = Unique PLC.Unique | Root deriving (Show, Eq, Ord)
+
+-- | A constraint requiring @g@ to be a 'G.Graph' (so we can compute e.g. a @Relation@ from it), whose
+-- vertices are 'Node's.
+type DepGraph g = (G.Graph g, (G.Vertex g)~Node)
+
+-- | Compute the dependency graph of a 'Term'. The 'Root' node will correspond to the term itself.
+runTermDeps
+    :: (DepGraph g, PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+    => Term tyname name a
+    -> g
+runTermDeps t = runReader (termDeps t) Root
+
+-- | Compute the dependency graph of a 'Type'. The 'Root' node will correspond to the type itself.
+runTypeDeps
+    :: (DepGraph g, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Type tyname a
+    -> g
+runTypeDeps t = runReader (typeDeps t) Root
+
+-- | Record some dependencies on the current node.
+recordDeps
+    :: (DepGraph g, MonadReader Node m)
```

Comment:
Why the use of `MonadReader Node`? Surely this could just take a `Node` as a parameter.